internal/planner/planner.go

```diff
@@ -2271,15 +2272,26 @@ func (p *Planner) optimizeLookup(t *ruletrie, ref ast.Ref) ([][]*ast.Rule, []ir.
 
 		nodes = nextNodes
 
-		// if all nodes have 0 children, abort ref check and optimize
-		all := true
+		// if all nodes have rules() > 0, abort ref check and optimize
+		// NOTE(sr): for a.b[c] = ... and a.b.d = ..., we stop at a.b, as its rules()
+		// will collect the children rules
+		// We keep the "all nodes have 0 children" check since it's cheaper and might
```

Comment:
When you mentioned a performance hit before (out-of-band), was the exclusion of this check part of that result?

Reply:
We're planning things differently now... And since we abort the optimization early, there's more to dot afterwards... But it was wrong before, so, that's probably ok.

There was a 12 second penalty at some point because I had included the node struct in a debug print, but that's been fixed 😁